query-planner-js/src/composedSchema/buildComposedSchema.ts

```diff
@@ -0,0 +1,240 @@
+    const name = graphValue.name;
+
+    const graphDirectiveArgs = getArgumentValuesForDirective(
+      graphDirective,
+      graphValue.astNode!,
+    );
+    assert(
+      graphDirectiveArgs,
+      `${graphEnumType.name} value ${name} in composed schema should have a @${graphDirective.name} directive`,
+    );
+
+    const serviceName: string = graphDirectiveArgs['name'];
+    const url: string = graphDirectiveArgs['url'];
+
+    graphMap[name] = {
+      serviceName,
+      url,
+    };
+  }
+
+  for (const type of Object.values(schema.getTypeMap())) {
+    if (isIntrospectionType(type)) continue;
+
+    // We currently only allow join spec directives on object types.
+    if (!isObjectType(type)) continue;
+
+    assert(
+      type.astNode,
+      `GraphQL type "${type.name}" should contain AST nodes`,
+    );
+
+    const ownerDirectiveArgs = getArgumentValuesForDirective(
+      ownerDirective,
+      type.astNode,
+    );
+
+    const typeMetadata: FederationTypeMetadata = ownerDirectiveArgs
+      ? {
+          serviceName: graphMap[ownerDirectiveArgs?.['graph']].serviceName,
+          keys: new MultiMap(),
+          isValueType: false,
+        }
+      : {
+          isValueType: true,
+        };
+
+    type.extensions = {
+      ...type.extensions,
+      federation: typeMetadata,
+    };
+
+    const typeDirectivesArgs = getArgumentValuesForRepeatableDirective(
+      typeDirective,
+      type.astNode,
+    );
+
+    assert(
+      !(typeMetadata.isValueType && typeDirectivesArgs.length >= 1),
+      `GraphQL type "${type.name}" cannot have a @${typeDirective.name} \
+directive without an $${ownerDirective.name} directive`,
+    );
+
+    for (const typeDirectiveArgs of typeDirectivesArgs) {
+      const serviceName = graphMap[typeDirectiveArgs['graph']].serviceName;
+
+      const keyFields = parseFieldSet(typeDirectiveArgs['key']);
+
+      typeMetadata.keys?.add(serviceName, keyFields);
+    }
+
+    for (const fieldDef of Object.values(type.getFields())) {
+      assert(
+        fieldDef.astNode,
+        `Field "${type.name}.${fieldDef.name}" should contain AST nodes`,
+      );
+
+      const fieldDirectiveArgs = getArgumentValuesForDirective(
+        fieldDirective,
+        fieldDef.astNode,
+      );
+
+      if (!fieldDirectiveArgs) continue;
+
+      const fieldMetadata: FederationFieldMetadata = {
+        serviceName: graphMap[fieldDirectiveArgs?.['graph']]?.serviceName,
+      };
+
+      fieldDef.extensions = {
+        ...fieldDef.extensions,
+        federation: fieldMetadata,
+      };
+
+      if (fieldDirectiveArgs) {
+        const { requires, provides } = fieldDirectiveArgs;
+
+        if (requires) {
+          fieldMetadata.requires = parseFieldSet(requires);
+        }
+
+        if (provides) {
+          fieldMetadata.provides = parseFieldSet(provides);
+        }
+      }
+    }
+  }
+
+  // We filter out schema elements that should not be exported to get to the
+  // API schema.
```

Comment:
This still leaves in the applications of directives, as discoverable via `astNode`? I think if we are filtering out the core/join stuff we should remove those too... certainly plenty of code (like graphql-tools-style schema directive processors) look at the astNodes.

eg this test fails

```
diff --git a/query-planner-js/src/composedSchema/__tests__/buildComposedSchema.test.ts b/query-planner-js/src/composedSchema/__tests__/buildComposedSchema.test.ts
index b0222786..91bbf3b7 100644
--- a/query-planner-js/src/composedSchema/__tests__/buildComposedSchema.test.ts
+++ b/query-planner-js/src/composedSchema/__tests__/buildComposedSchema.test.ts
@@ -36,6 +36,11 @@ describe('buildComposedSchema', () => {
     expect(types).toEqual([]);
   });
 
+  it(`removes core directive applications from the AST`, () => {
+    const directives = schema.astNode?.directives;
+    expect(directives).toEqual([]);
+  });
+
   it(`doesn't include join directives`, () => {
     const directives = schema
       .getDirectives()
```

The easiest way to implement this might honestly be to print -> parse -> [use a visitor like this `exportSchema` from core-schema-js](https://github.com/apollographql/core-schema-js/pull/2/files#diff-864741d7c69b18cce20ffe48a7e72c0670ccfb5c391d6ff9d26742ce93a50d8dR98-R105) -> buildASTSchema again...